### ISSUE TITLE
some bug 

### ISSUE THREAD

**Issue by rtfingc** [enferex/pdfresurrect#6]
I find some bugs through fuzz. You can check it out.

```c
==52118==WARNING: AddressSanitizer failed to allocate 0xffffffffffffffe8 bytes
==52118==AddressSanitizer's allocator is terminating the process instead of returning 0
==52118==If you don't like this behavior set allocator_may_return_null=1
==52118==AddressSanitizer CHECK failed: /build/llvm-toolchain-6.0-QjOn7h/llvm-toolchain-6.0-6.0/projects/compiler-rt/lib/sanitizer_common/sanitizer_allocator.cc:225 "((0)) != (0)" (0x0, 0x0)
    #0 0x4e3165 in __asan::AsanCheckFailed(char const*, int, char const*, unsigned long long, unsigned long long) (/home/rtfingc/test/pdfresurrect2/pdfresurrect+0x4e3165)
    #1 0x500a15 in __sanitizer::CheckFailed(char const*, int, char const*, unsigned long long, unsigned long long) (/home/rtfingc/test/pdfresurrect2/pdfresurrect+0x500a15)
    #2 0x4e9556 in __sanitizer::ReportAllocatorCannotReturnNull() (/home/rtfingc/test/pdfresurrect2/pdfresurrect+0x4e9556)
    #3 0x4e9596 in __sanitizer::ReturnNullOrDieOnFailure::OnBadRequest() (/home/rtfingc/test/pdfresurrect2/pdfresurrect+0x4e9596)
    #4 0x425179 in __asan::asan_calloc(unsigned long, unsigned long, __sanitizer::BufferedStackTrace*) (/home/rtfingc/test/pdfresurrect2/pdfresurrect+0x425179)
    #5 0x4da212 in calloc (/home/rtfingc/test/pdfresurrect2/pdfresurrect+0x4da212)
    #6 0x51c061 in load_xref_from_plaintext /home/rtfingc/test/pdfresurrect2/pdf.c:646:21
    #7 0x5162c1 in load_xref_entries /home/rtfingc/test/pdfresurrect2/pdf.c:623:7
    #8 0x5155c1 in pdf_load_xrefs /home/rtfingc/test/pdfresurrect2/pdf.c:279:9
    #9 0x5130a6 in init_pdf /home/rtfingc/test/pdfresurrect2/main.c:206:9
    #10 0x512a24 in main /home/rtfingc/test/pdfresurrect2/main.c:263:17
    #11 0x7fdcd3b5cb96 in __libc_start_main /build/glibc-OTsEL5/glibc-2.27/csu/../csu/libc-start.c:310
    #12 0x41a159 in _start (/home/rtfingc/test/pdfresurrect2/pdfresurrect+0x41a159)
```
in function ```load_xref_from_plaintext```
```c
    SAFE_E(fread(buf, 1, 21, fp), 21, "Failed to load entry Size string.\n");
    xref->n_entries = atoi(buf + strlen("ize "));
    xref->entries = calloc(1, xref->n_entries * sizeof(struct _xref_entry));
```
buf can be control, and then we can control n_entries without any check

---

```c
==52201==WARNING: AddressSanitizer failed to allocate 0xfffffffffffffda4 bytes
==52201==AddressSanitizer's allocator is terminating the process instead of returning 0
==52201==If you don't like this behavior set allocator_may_return_null=1
==52201==AddressSanitizer CHECK failed: /build/llvm-toolchain-6.0-QjOn7h/llvm-toolchain-6.0-6.0/projects/compiler-rt/lib/sanitizer_common/sanitizer_allocator.cc:225 "((0)) != (0)" (0x0, 0x0)
    #0 0x4e3165 in __asan::AsanCheckFailed(char const*, int, char const*, unsigned long long, unsigned long long) (/home/rtfingc/test/pdfresurrect2/pdfresurrect+0x4e3165)
    #1 0x500a15 in __sanitizer::CheckFailed(char const*, int, char const*, unsigned long long, unsigned long long) (/home/rtfingc/test/pdfresurrect2/pdfresurrect+0x500a15)
    #2 0x4e9556 in __sanitizer::ReportAllocatorCannotReturnNull() (/home/rtfingc/test/pdfresurrect2/pdfresurrect+0x4e9556)
    #3 0x4e9596 in __sanitizer::ReturnNullOrDieOnFailure::OnBadRequest() (/home/rtfingc/test/pdfresurrect2/pdfresurrect+0x4e9596)
    #4 0x425116 in __asan::asan_malloc(unsigned long, __sanitizer::BufferedStackTrace*) (/home/rtfingc/test/pdfresurrect2/pdfresurrect+0x425116)
    #5 0x4d9feb in malloc (/home/rtfingc/test/pdfresurrect2/pdfresurrect+0x4d9feb)
    #6 0x517ce0 in pdf_load_pages_kids /home/rtfingc/test/pdfresurrect2/pdf.c:317:19
    #7 0x5130d9 in init_pdf /home/rtfingc/test/pdfresurrect2/main.c:210:5
    #8 0x512a24 in main /home/rtfingc/test/pdfresurrect2/main.c:263:17
    #9 0x7fb2fa0d7b96 in __libc_start_main /build/glibc-OTsEL5/glibc-2.27/csu/../csu/libc-start.c:310
    #10 0x41a159 in _start (/home/rtfingc/test/pdfresurrect2/pdfresurrect+0x41a159)
```
in pdf_load_pages_kids, another problem similar to the firsrt on.
```c
            sz = pdf->xrefs[i].end - ftell(fp);
            buf = malloc(sz + 1);
            SAFE_E(fread(buf, 1, sz, fp), sz, "Failed to load /Root.\n");
            buf[sz] = '\0';
```

---
```c
memleak2: --D-- Version 2 -- Object 70 (Unknown)
memleak2: --A-- Version 2 -- Object 71 (Unknown)
memleak2: --D-- Version 3 -- Object 70 (Unknown)
memleak2: --?-- Version 3 -- Object 71 (Unknown)
---------- memleak2 ----------
Versions: 3

=================================================================
==52214==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 1280 byte(s) in 1 object(s) allocated from:
    #0 0x4da490 in realloc (/home/rtfingc/test/pdfresurrect2/pdfresurrect+0x4da490)
    #1 0x518443 in get_object /home/rtfingc/test/pdfresurrect2/pdf.c:1115:18
    #2 0x51b955 in get_object_from_here /home/rtfingc/test/pdfresurrect2/pdf.c:1058:12
    #3 0x5160d2 in is_valid_xref /home/rtfingc/test/pdfresurrect2/pdf.c:603:13
    #4 0x5153b8 in pdf_load_xrefs /home/rtfingc/test/pdfresurrect2/pdf.c:268:14
    #5 0x5130a6 in init_pdf /home/rtfingc/test/pdfresurrect2/main.c:206:9
    #6 0x512a24 in main /home/rtfingc/test/pdfresurrect2/main.c:263:17
    #7 0x7fac10544b96 in __libc_start_main /build/glibc-OTsEL5/glibc-2.27/csu/../csu/libc-start.c:310

SUMMARY: AddressSanitizer: 1280 byte(s) leaked in 1 allocation(s).

```
memleak

---
```c
=================================================================
==52243==ERROR: AddressSanitizer: stack-buffer-overflow on address 0x7ffd169be7e0 at pc 0x00000051747e bp 0x7ffd169be790 sp 0x7ffd169be788
WRITE of size 1 at 0x7ffd169be7e0 thread T0
    #0 0x51747d in load_creator /home/rtfingc/test/pdfresurrect2/pdf.c:877:33
    #1 0x51567a in pdf_load_xrefs /home/rtfingc/test/pdfresurrect2/pdf.c:291:5
    #2 0x5130a6 in init_pdf /home/rtfingc/test/pdfresurrect2/main.c:206:9
    #3 0x512a24 in main /home/rtfingc/test/pdfresurrect2/main.c:263:17
    #4 0x7f8cbeb2db96 in __libc_start_main /build/glibc-OTsEL5/glibc-2.27/csu/../csu/libc-start.c:310
    #5 0x41a159 in _start (/home/rtfingc/test/pdfresurrect2/pdfresurrect+0x41a159)
                 
Address 0x7ffd169be7e0 is located in stack of thread T0 at offset 64 in frame
    #0 0x516a8f in load_creator /home/rtfingc/test/pdfresurrect2/pdf.c:832

  This frame has 2 object(s):
    [32, 64) 'obj_id_buf' (line 834) <== Memory access at offset 64 overflows this variable
    [96, 104) 'sz' (line 836)
HINT: this may be a false positive if your program uses some custom stack unwind mechanism or swapcontext
      (longjmp and C++ exceptions *are* supported)
SUMMARY: AddressSanitizer: stack-buffer-overflow /home/rtfingc/test/pdfresurrect2/pdf.c:877:33 in load_creator
Shadow bytes around the buggy address:
  0x100022d2fca0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x100022d2fcb0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x100022d2fcc0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x100022d2fcd0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x100022d2fce0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
=>0x100022d2fcf0: 00 00 00 00 f1 f1 f1 f1 00 00 00 00[f2]f2 f2 f2

```
stackoverflow

>

**Comment by rtfingc** [comment]
[stackoverflow_some.pdf](https://github.com/enferex/pdfresurrect/files/3480367/stackoverflow_some.pdf)
[stackoverflow2.pdf](https://github.com/enferex/pdfresurrect/files/3480368/stackoverflow2.pdf)
[calloc_some.pdf](https://github.com/enferex/pdfresurrect/files/3480369/calloc_some.pdf)
[malloc_some.pdf](https://github.com/enferex/pdfresurrect/files/3480370/malloc_some.pdf)
[memleak2.pdf](https://github.com/enferex/pdfresurrect/files/3480372/memleak2.pdf)
[memory_leak.pdf](https://github.com/enferex/pdfresurrect/files/3480373/memory_leak.pdf)

**Comment by enferex** [comment]
Fixed as part of v0.18
